components/browser/menu/src/main/java/mozilla/components/browser/menu/view/ExpandableLayout.kt

```diff
@@ -0,0 +1,379 @@
+                MotionEvent.ACTION_DOWN -> {
+                    if (isExpandInProgress) {
+                        return true
+                    }
+
+                    // Check if user clicked in the empty space left by this collapsed View.
+                    if (!isTouchingTheWrappedView(ev)) {
+                        blankTouchListener?.invoke()
+                    }
+
+                    initialYCoord = ev.y
+
+                    false // Allow click listeners firing for children.
+                }
+                MotionEvent.ACTION_MOVE -> {
+                    if (isExpandInProgress || !isCollapsed) {
+                        return true
+                    }
+
+                    if (isScrollingUp(ev)) {
+                        expand()
+                        true
+                    } else {
+                        false
+                    }
+                }
+                else -> {
+                    // In general, we don't want to intercept touch events.
+                    // They should be handled by the child view.
+                    return callParentOnInterceptTouchEvent(ev)
+                }
+            }
+        }
+
+        return callParentOnInterceptTouchEvent(ev)
+    }
+
+    @VisibleForTesting
+    internal fun shouldInterceptTouches() = isCollapsed || parentHeight > expandedHeight
+
+    @VisibleForTesting
+    internal fun isTouchingTheWrappedView(ev: MotionEvent): Boolean {
+        val childrenBounds = Rect()
+        wrappedView.getHitRect(childrenBounds)
+        return childrenBounds.contains(ev.x.toInt(), ev.y.toInt())
+    }
+
+    @VisibleForTesting
+    internal fun collapse() {
+        wrappedView.translationY = parentHeight.toFloat() - collapsedHeight
+        wrappedView.updateLayoutParams {
+            height = collapsedHeight
+        }
+    }
+
+    @VisibleForTesting
+    internal fun expand() {
+        isCollapsed = false
+        isExpandInProgress = true
+
+        val initialTranslation = wrappedView.translationY
+        val distanceToExpandedHeight = expandedHeight - collapsedHeight
+        getExpandViewAnimator(distanceToExpandedHeight).apply {
+            doOnEnd {
+                isExpandInProgress = false
+            }
+
+            addUpdateListener {
+                wrappedView.translationY = initialTranslation - it.animatedValue as Int
+                wrappedView.updateLayoutParams {
+                    height = collapsedHeight + it.animatedValue as Int
+                }
+            }
+            start()
+        }
+    }
+
+    @VisibleForTesting
+    internal fun getExpandViewAnimator(expandDelta: Int): ValueAnimator {
+        return ValueAnimator.ofInt(0, expandDelta).apply {
+            this.interpolator = AccelerateDecelerateInterpolator()
+            this.duration = DEFAULT_DURATION_EXPAND_ANIMATOR
+        }
+    }
+
+    @VisibleForTesting
+    internal fun getOrCalculateCollapsedHeight(): Int {
+        // Memoize the value.
```

Comment:
nit: Memorize?

Reply:
Used "memoize" to express a bit more about the motivation for caching it - to avoid repeating future expensive operations.
https://en.wikipedia.org/wiki/Memoization

Reply:
Nice 🤓 !